### PR TITLE
feat: (Store) Entity format

### PR DESCRIPTION
**Pull request by genereg1**
#### Please provide a link to the associated issue.
https://github.com/SAP/fundamental-ngx/issues/4583
#### Please provide a brief summary of this pull request.

- Basic implementation BaseEntity and BaseValue (Value Object)
- New store method `createEntityInstance` 
- Proxy instance wrapper to treat Entity properties (nested properties e.g arrays)

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
- [x] Run npm run build-pack-library and test in external application

Documentation checklist:
- [x] update `README.md`
- [x ] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

### CODE REVIEW

File: libs/store/src/lib/infrastructure/persistence/domain/base-classes/base-entity.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import { Composite } from './composite';
+
+/**
+ * BaseEntityDTO that extends passed Entity DTO
+ * */
+export interface BaseEntityDTO {
+    id: number;
```
Comment (KevinOkamoto) [inline libs/store/src/lib/infrastructure/persistence/domain/base-classes/base-entity.ts:7]: Are we expecting the primary key of a DTO to always be named `id` and be of type `number`? In many cases the primary key will be named something else and be of type `string`.

I think we allow the user to set what the primary key is in the Entity decorator:
https://github.com/SAP/fundamental-ngx/blob/ab0dad82f415e3fc2150c5f45b507c6c9d01276d/libs/store/src/lib/domain/entity.ts#L11

File: libs/store/src/lib/infrastructure/persistence/store/entity-store.ts [hunk 2]
```diff
@@ -90,7 +99,7 @@ export class DefaultQueryService<TModel> extends QueryService<TModel> {
 
     getWithQuery(query: QuerySnapshot<TModel>): Observable<TModel[]> {
         // Pass query payload instead of QueryParams
-        // TODO: Should we override EntityCollectionService interface for that? 
+        // TODO: Should we override EntityCollectionService interface for that?
         return this.service.getWithQuery(query as any);
```
Comment (KevinOkamoto) [inline libs/store/src/lib/infrastructure/persistence/store/entity-store.ts:103]: We need to map the return data here (and within the `getByKey` method) from the EntityCollectionService to call `instanceForType`, so that the end-user interacts with Entity objects and not DTOs.

**Review by KevinOkamoto — CHANGES_REQUESTED** [review]

File: libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    get(target: any, prop: PropertyKey): any {
+        // Check is an array and wraps a new one proxy
+        if (['[object Object]', '[object Array]']
+            .indexOf(Object.prototype.toString.call(this.targetState[prop])) > -1) {
+            return new Proxy(this.targetState[prop], this);
+        }
+        return target[prop] || this.targetState[prop];
+    }
+
+    set<T>(target: TModel, key: PropertyKey, value: Type<T> | Array<Type<T>>): boolean {
```
Comment (dimamarksman) [inline libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts:28]: As for me, there is no much sense in `T` in this case.
Let's handle `value` as `unknown` and use `instanceof` to guess what it actually is.

File: libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    constructor(
+        private targetState?: TModel
+    ) {
+    }
+
+    get(target: any, prop: PropertyKey): any {
+        // Check is an array and wraps a new one proxy
+        if (['[object Object]', '[object Array]']
+            .indexOf(Object.prototype.toString.call(this.targetState[prop])) > -1) {
+            return new Proxy(this.targetState[prop], this);
```
Comment (dimamarksman) [inline libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts:23]: That means that every read operation will create a new instance of proxy, Am I right?
For example:
```
read entityProxy.lineItems ---> create new instance of Proxy
...
read entityProxy.lineItems ---> create new instance of Proxy 
```

I wonder if we could create only one Proxy instance for each property that needs to be wrapped. Can we do that?

File: libs/store/src/lib/infrastructure/persistence/store/entity-store-builder.ts [hunk 5]
```diff
@@ -42,7 +42,7 @@ export abstract class EntityStoreBuilderFactory {
 /**
  * Entity Store Builder default implementation
  */
-export class DefaultEntityStoreBuilder<T extends BaseEntity> implements EntityStoreBuilder<T> {
```
Comment (dimamarksman) [inline libs/store/src/lib/infrastructure/persistence/store/entity-store-builder.ts:45]: Why did you remove `extends BaseEntity` constraint?

File: libs/store/src/lib/infrastructure/persistence/store/entity-store.ts [hunk 6]
```diff
@@ -33,6 +34,12 @@ export interface EntityStore<T> {
      * @param entity Entity to be deleted
      */
     delete(entity: T): Observable<T>;
+
+    /**
+     * Create new entity instance wrapped Proxy
+     * @param fromState initial state
+     */
+    createEntityInstance(fromState?: any): T;
```
Comment (dimamarksman) [inline libs/store/src/lib/infrastructure/persistence/store/entity-store.ts:42]: can we avoid `any`?

**Review by dimamarksman — CHANGES_REQUESTED** [review]
Have got some question, please take a look

File: libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts (hunk 4 above)
Comment (dimamarksman) [inline libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts:23]: Can we really share the same Proxy handler with the new child Proxy?
Asking because `this` handler shares the same `targetState` for all underlying proxies. 
At the same time we use `this.targetState` to make an assignment for each particular proxy which represents different model shapes. 
Does it really work? Just can't get it...

File: libs/store/src/lib/infrastructure/persistence/domain/base-classes/base-entity.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Base Entity follows the DDD definition of Entity
+ */
+export abstract class BaseEntity<EntityProps extends BaseEntityDTO> extends Composite<EntityProps> {
+     constructor(dto: EntityProps) {
+         super();
+         this._dto = dto;
+     }
+     _dto: EntityProps;
```
[the remark this answers is not in the document]
Comment (genereg1) [reply libs/store/src/lib/infrastructure/persistence/domain/base-classes/base-entity.ts:18]: `this._dto` used by `equals` method in `BaseEntity` class just to check id equality. Probably, a better way to assign just only id property `this._id = dto.id`

File: libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts (hunk 4 above)
Comment (genereg1) [reply libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts:23]: Good point. I will check what we can do with it

File: libs/store/src/lib/infrastructure/persistence/domain/base-classes/composite.ts [hunk 8]
```diff
@@ -0,0 +1,4 @@
+export abstract class Composite<T> {
+    value: T;
```
[the remark this answers is not in the document]
Comment (genereg1) [reply libs/store/src/lib/infrastructure/persistence/domain/base-classes/composite.ts:2]: Actually, it doesn't use. Initially, the idea was to create some interface/class to get access to proxy value or get errors from there. But currently, I don't see a case when we could use it.

File: libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        }
+        return target[prop] || this.targetState[prop];
```
[the remark this answers is not in the document]
Comment (genereg1) [reply libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts:25]: Yea. I think it's a better way to do it

File: libs/store/src/lib/infrastructure/persistence/domain/base-classes/base-value.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    protected constructor(dto: T) {
+        super();
+        this.dto = Object.freeze(dto);
+    }
+
+    /**
+     * Clone Value Object since they are immutable
+     */
+    clone() {
+        return { ...this.dto };
```
[the remark this answers is not in the document]
Comment (genereg1) [reply libs/store/src/lib/infrastructure/persistence/domain/base-classes/base-value.ts:15]: From my perspective, the `clone` method will be used in the case when we want to edit our VO. Currently, the method just creates a copy of non-editable dto that would be editable, but the original dto will be still non-editable.

File: libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    set<T>(target: TModel, key: PropertyKey, value: Type<T> | Array<Type<T>>): boolean {
+        if (value instanceof BaseValue) {
+            if (!value.dto) {
+                this.targetState[key] = value;
+                return true;
+            }
+            const isEqual = Array.isArray(target) && target.some(elem => value.equals(elem)
+                || value.equals(target[key]));
+
+            let clonedVO = target[key] && value.clone();
```
[the remark this answers is not in the document]
Comment (genereg1) [reply libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts]: Removed

File: libs/store/src/lib/infrastructure/persistence/domain/base-classes/base-value.ts [hunk 12]
```diff
@@ -0,0 +1,28 @@
+import { Composite } from './composite';
+
+export abstract class BaseValue<T> extends Composite<T>{
+    readonly dto: T;
+
+    protected constructor(dto: T) {
+        super();
+        this.dto = Object.freeze(dto);
```
[the remark this answers is not in the document]
Comment (genereg1) [reply libs/store/src/lib/infrastructure/persistence/domain/base-classes/base-value.ts]: Added deepFreeze function

File: libs/store/src/lib/infrastructure/persistence/domain/base-classes/base-entity.ts [hunk 13]
```diff
@@ -0,0 +1,49 @@
+import { Composite } from './composite';
+
+/**
+ * BaseEntityDTO that extends passed Entity DTO
+ * */
+export interface BaseEntityDTO {
+    id: number;
+    primaryKey?: string;
```
Comment (KevinOkamoto) [inline libs/store/src/lib/infrastructure/persistence/domain/base-classes/base-entity.ts:8]: You don't need to add the `primaryKey` property here. What I was trying to say was that the DTO from a service might not have `id` as its identifier, it may be some other field, like:
```ts
{
   uuid: '1a3dd8b8-bb2a-4e62-92f7-6f5f5cb2ea7' ,
   name: 'John',
   age: 24
}
```
Here `uuid` is the identifier for the DTO, and it is a string.

File: libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts [hunk 14]
```diff
@@ -34,15 +33,12 @@ class StateHandler<TModel> {
...
+                target[key] = new (Object.getPrototypeOf(value).constructor)(value.dto);
                 return true;
             }
         } else {
-            this.targetState[key] = value;
-            return true;
+            return Reflect.set(this.targetState as any, key, value);
         }
 
         return true;
```
Comment (dimamarksman) [inline libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts:44]: Should be `false` probably

File: apps/store-app/src/app/entities/requisition.entity.ts [hunk 15]
```diff
@@ -0,0 +1,27 @@
+import 'reflect-metadata';
```
Comment (fkolar) [inline apps/store-app/src/app/entities/requisition.entity.ts:1]: I have already commented something similar in other PR but angular is getting rid of this polyfils.  Should not we do the same? 

https://angular.io/guide/deprecations#reflect-metadata

File: libs/store/src/lib/infrastructure/persistence/domain/base-classes/base-entity.ts (hunk 1 above)
Comment (fkolar) [reply libs/store/src/lib/infrastructure/persistence/domain/base-classes/base-entity.ts:7]: There could be variety of ways to specify which key is primary key. In ariba especially we used many times. This is the reason we started originally with `Identity`  so we can describe what field and type. as kevin says we need to be able to specify which field is primary key

File: libs/store/src/lib/infrastructure/persistence/domain/base-classes/composite.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (fkolar) [reply libs/store/src/lib/infrastructure/persistence/domain/base-classes/composite.ts:2]: Its either Composite<T> or our BaseEntity/BaseValue. It's the same, both have a internal DTO state and we read values from it or set value. so composite supposed to be abstract class that BaseEntity/BaseValue inherit from as both have internal DTO State. 

and Yes original idea was to handle also the NULL/undefined problem (the articles we posted into slack channel). + expose some kind of xx$ state to this entity level . 

So the `value: T` is our `readonly dto: T` it is duplicate information.

File: libs/store/src/lib/infrastructure/persistence/domain/base-classes/base-value.ts [hunk 16]
```diff
@@ -0,0 +1,45 @@
+import { Composite } from './composite';
+
+export abstract class BaseValue<T> extends Composite<T>{
+    readonly dto: T;
```
Comment (fkolar) [inline libs/store/src/lib/infrastructure/persistence/domain/base-classes/base-value.ts:4]: see comment bellow about `Composite`

**Review by fkolar — CHANGES_REQUESTED** [review]
Just a few minor comments most of the things were covered already

File: libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts (hunk 4 above)
Comment (fkolar) [reply libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts]: I am not sure if this can be the same instance, as proxy does not handle dotted notation, of the Target, key, value is always for the main object

File: libs/store/src/lib/infrastructure/persistence/domain/state-handler.spec.ts [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+
+interface ReqDTO extends BaseEntityDTO {
+    id: number;
+    title: string;
+    amount: number;
+    lineItems: Array<LineItem>;
+}
+
+class Requisition extends BaseEntity<ReqDTO> {
+    id: number;
```
Comment (fkolar) [inline libs/store/src/lib/infrastructure/persistence/domain/state-handler.spec.ts:51]: How we ended up with the discussion about the ID?  In Ariba I see we can have :

```
class Requisition {uniqueName: string}

class SourcingEvent {internalId: string;}

class SomeOtherEntity {id: string)

Now we need to have way to point to the correct key. 

What if our baseEntity will have method/getter/setter. 

identity(): string, so that in the Requisition about you will implement it and call the right DTO attribute? 

```
class Requisition extends BaseEntity {
   identity(): string {
    return  this.value.id;
   }
 
}
```

or maybe just getter/setter

**Review by fkolar — CHANGES_REQUESTED** [review]

File: libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+            const isEqual = Array.isArray(target) && target.some(elem => value.equals(elem)
+                || value.equals(target[key]));
+
+            if (!isEqual) {
+                // target[key] = value;
+                // return true;
+                return Reflect.set(target as any, key, value);
+            }
+        } else {
+            return Reflect.set(this.state as any, key, value);
```
Comment (dimamarksman) [inline libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts:40]: Why do we use `this.state` as a target in this case?

Comment (dimamarksman) [reply libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts:40]:
```
it('Test nesting', () => {
        class Child {
            title: string;
        }
        
        interface DTO {
            id: string;
            title: string;
            child: Child;
        }
        class Primary extends BaseEntity<DTO> {
            title: string;
            child: Child;
        }

        const initState: DTO = {
            id: '123',
            title: 'Primary',
            child: {
                title: 'Child'
            }
        };

        const primary: Primary = instanceForType(Primary, initState);

        expect(primary.title).toBe('Primary');

        primary.child.title = 'New Child';

        expect(primary.title).toBe('Primary'); --> will fail cause 'primary.title' will be 'New Child'
    });
```

**Review by dimamarksman — CHANGES_REQUESTED** [review]

File: libs/store/src/lib/domain/decorators.ts [hunk 19]
```diff
@@ -1,40 +1,44 @@
...
 const ENTITY_META_MAP = new Map<Type<any>, EntityMetaOptions<any>>();
 const RESOURCE_MAP = new Map<Type<any>, EntityResourceMetaOptions>();
 
-export type EntityType<T extends BaseEntity> = Type<T>;
+export type EntityType<T> = Type<T>;
 
-export function Entity<T extends BaseEntity>(config: EntityMetaOptions<T>) {
+export interface EntityClass<T> extends BaseEntity<T> {
+    new (...arg: any[]): EntityClass<T>;
+}
```
[the remark this answers is not in the document]
Comment (dimamarksman) [reply libs/store/src/lib/domain/decorators.ts:16]: Instead you might need something like that:
```
export type EntityClass<E extends BaseEntity<K>, K = any> = new (...args: any[]) => E;
```

**Review by dimamarksman — CHANGES_REQUESTED** [review]

File: libs/store/src/lib/infrastructure/persistence/store/entity-store-builder.spec.ts [hunk 20]
```diff
@@ -5,37 +5,47 @@ import {
...
-    getEntityResourceMetadata<T extends BaseEntity>(entity: string | EntityType<T>): EntityResourceMetaOptions {
+    getEntityResourceMetadata<T>(entity: string | EntityType<T>): EntityResourceMetaOptions {
         throw new Error('Method not implemented.');
     }
-    getEntityMetadata<T extends BaseEntity>(entity: string | EntityType<T>): EntityMetaOptions<T> {
+    getEntityMetadata<T>(entity: string | EntityType<T>): EntityMetaOptions<T> {
+        throw new Error('Method not implemented.');
+    }
+
+    getEntityTypeByName(entityName: string): EntityBaseType {
```
Comment (dimamarksman) [inline libs/store/src/lib/infrastructure/persistence/store/entity-store-builder.spec.ts:39]: return type should be `EntityClass` probably

File: libs/store/src/lib/infrastructure/persistence/store/entity-store-builder.ts [hunk 21]
```diff
@@ -46,6 +46,7 @@ export interface EntityStoreBuilder<T extends BaseEntity> {
  * Entity Store Builder Factory interface
  */
 export abstract class EntityStoreBuilderFactory {
+    // abstract create<T extends BaseEntity<T>>(entity: EntityType<T>): EntityStoreBuilder<T>;
```
Comment (dimamarksman) [inline libs/store/src/lib/infrastructure/persistence/store/entity-store-builder.ts:49]: please remove this comment

File: libs/store/src/lib/infrastructure/persistence/store/entity-store.ts [hunk 22]
```diff
@@ -37,13 +39,19 @@ export interface EntityStore<T> {
      * @param entity Entity to be deleted
      */
     delete(entity: T): Observable<T>;
+
+    /**
+     * Create new entity instance wrapped Proxy
+     * @param fromState initial state
+     */
+    createEntityInstance(fromState?: T): T;
```
Comment (dimamarksman) [inline libs/store/src/lib/infrastructure/persistence/store/entity-store.ts:47]: Actually `fromState` should be based on EntityDTO generic type, `T` in this context means entity class type, not a DTO type

File: libs/store/src/lib/infrastructure/persistence/store/entity-store.spec.ts [hunk 23]
```diff
@@ -37,14 +45,15 @@ class UserCollectionServiceMock implements Partial<EntityCollectionService<User>
 class QueryBuilderMock extends QueryBuilder<User> {}
 
 describe('Default Entity Store', () => {
+    let entity: Type<User>;
```
Comment (dimamarksman) [inline libs/store/src/lib/infrastructure/persistence/store/entity-store.spec.ts:48]: We have to cover `createEntityInstance` method by unit tests

**Review by dimamarksman — CHANGES_REQUESTED** [review]

File: libs/store/src/lib/infrastructure/persistence/store/entity-store.ts [hunk 24]
```diff
@@ -70,7 +79,7 @@ export class DefaultEntityStore<T extends BaseEntity> implements EntityStore<T>
     }
```
Comment (KevinOkamoto) [inline libs/store/src/lib/infrastructure/persistence/store/entity-store.ts:79]: When getting an entity from the EntityService here:
```ts
    get(id: IdentityKey): Observable<T> {
        return this._entityService.getByKey(id);
    }
```

Does the EntityService return an Observable with the Entity or EntityDTO?

Shouldn't we be calling `instanceForType` here? For example:

```ts
    get(id: IdentityKey): Observable<T> {
        return this._entityService.getByKey(id).pipe(
            map(dto: EntityDTO => instanceForType(this._entity, dto))
        );
    }
```

Comment (fkolar) [reply libs/store/src/lib/infrastructure/persistence/store/entity-store.ts:79]: I think we should be returning Entity, EntityDTO is an internal thing

File: libs/store/src/lib/domain/decorators.ts [hunk 25]
```diff
@@ -1,40 +1,44 @@
-import { BaseEntity } from './entity';
 import { EntityMetaOptions } from './entity-meta-options';
 import { EntityResourceMetaOptions } from './rest-resource';
 import { Type } from './utility';
+import { BaseEntity } from '../infrastructure/persistence';
```
[the remark this answers is not in the document]
Comment (fkolar) [reply libs/store/src/lib/domain/decorators.ts:4]: I agree here.

File: libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+    ) {
+    }
+
+    get(target: any, prop: PropertyKey): any {
+        // Check is nested object and wraps a new one proxy
+        if (this.state[prop] && typeof this.state[prop] === 'object') {
+            return new Proxy(this.state[prop], this);
+        }
+
+        return Reflect.get(target as any, prop);
```
Comment (fkolar) [inline libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts:26]: Did not we say that we are not going to use Reflect? and Dima removed it from his PR as well?

File: libs/store/src/lib/infrastructure/persistence/query/query.ts [hunk 27]
```diff
@@ -1,5 +1,5 @@
 import { Observable } from 'rxjs';
-import { BaseEntity, ChainingStrategyFieldsMap } from '../../../domain/public_api';
+import { ChainingStrategyFieldsMap } from '../../../domain/public_api';
```
Comment (fkolar) [inline libs/store/src/lib/infrastructure/persistence/query/query.ts:2]: Let's use direct import not from barrel files or these public API files.

File: libs/store/src/lib/infrastructure/persistence/store/entity-collection-service-base.spec.ts [hunk 28]
```diff
@@ -6,7 +6,7 @@ import { EntityMetaOptions, EntityMetaOptionsService } from '../utils/entity-opt
 import { EntityCollectionService } from './entity-collection-service';
 import { DefaultEntityCollectionService } from './entity-collection-service-base';
 import { EntityCollectionsService } from './entity-collections-service';
-import { BaseEntity } from './entity-server/interfaces';
+import { BaseEntity, IdentityKey } from '@fundamental-ngx/store';
```
Comment (fkolar) [inline libs/store/src/lib/infrastructure/persistence/store/entity-collection-service-base.spec.ts:9]: for local test you should not be importing it from @xxxx/store

File: libs/store/src/lib/infrastructure/persistence/store/entity-store.spec.ts [hunk 29]
```diff
@@ -2,10 +2,18 @@ import { of } from 'rxjs';
 
 import { DefaultEntityStore } from './entity-store';
 import { QueryBuilder } from '../query/query-builder';
+import { IdentityKey, Type } from '../../../domain/public_api';
+import { BaseEntity } from '../domain/base-classes/base-entity';
```
Comment (fkolar) [inline libs/store/src/lib/infrastructure/persistence/store/entity-store.spec.ts:6]: as above , use direct imports

**Review by fkolar — CHANGES_REQUESTED** [review]

File: libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts (hunk 26 above)
Comment (dimamarksman) [reply libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts:26]: Maybe I missed something but why we shouldn't use Reflect?

Comment (dimamarksman) [reply libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts:26]: The only thing the Reflect has to use accessor option otherwise there is no much sense in it

File: libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts [hunk 30]
```diff
@@ -0,0 +1,47 @@
+import 'reflect-metadata';
```
Comment (dimamarksman) [inline libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts:1]: We don't need `reflect-metadata`

**Review by dimamarksman — CHANGES_REQUESTED** [review]

File: libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts (hunk 26 above)
Comment (dimamarksman) [reply libs/store/src/lib/infrastructure/persistence/domain/state-handler.ts:25]: @fkolar it's a native Reflect, not that one that comes from `reflect-metadata`

**Review by dimamarksman — APPROVED** [review]

**Review by KevinOkamoto — APPROVED** [review]

**Review by fkolar — APPROVED** [review]
